src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py

```diff
@@ -0,0 +1,365 @@
+        super().__init__()
+        self.config = config
+        self.layer_idx = layer_idx
+        self.moe_topk = config.moe_topk if isinstance(config.moe_topk, int) else config.moe_topk[layer_idx]
+        self.drop_tokens = config.moe_drop_tokens
+        self.random_routing_dropped_token = config.moe_random_routing_dropped_token
+        num_experts = config.num_experts if isinstance(config.num_experts, int) else config.num_experts[layer_idx]
+        self.wg = nn.Linear(config.hidden_size, num_experts, bias=False, dtype=torch.float32)
+
+    def forward(self, hidden_states):
+        bsz, seq_len, hidden_size = hidden_states.shape
+        hidden_states = hidden_states.reshape(-1, hidden_size)
+        if self.wg.weight.dtype == torch.float32:
+            hidden_states = hidden_states.float()
+        logits = self.wg(hidden_states)
+        gate_output = topkgating(logits, self.moe_topk)
+
+        return gate_output
+
+
+class HunYuanMoE(nn.Module):
+    def __init__(self, config: HunYuanMoEV1Config, layer_idx: Optional[int] = None):
+        super().__init__()
+        self.config = config
+        self.layer_idx = layer_idx
+        self.moe_topk = config.moe_topk
+        self.num_experts = config.num_experts if isinstance(config.num_experts, int) else config.num_experts[layer_idx]
+        self.shared_mlp = HunYuanMoEV1MLP(config, layer_idx=layer_idx, is_shared_mlp=True)
+        self.gate = HunYuanTopKGate(config, layer_idx=layer_idx)
+        self.experts = nn.ModuleList(
+            [HunYuanMoEV1MLP(config, layer_idx=layer_idx, is_shared_mlp=False) for _ in range(self.num_experts)]
+        )
+
+    def forward(self, hidden_states):
+        bsz, seq_len, hidden_size = hidden_states.shape
+
+        hidden_states_mlp = self.shared_mlp(hidden_states)
+
+        combine_weights, dispatch_mask = self.gate(hidden_states)
+
+        reshaped_input = hidden_states.reshape(-1, hidden_size)
+
+        # dispatched_input = torch.einsum("sec,sm->ecm", dispatch_mask.type_as(hidden_states), reshaped_input)
+        dispatch_mask_expanded = dispatch_mask.type_as(hidden_states).unsqueeze(3)  #  (s, e, c, 1)
+        reshaped_input_expanded = reshaped_input.unsqueeze(1).unsqueeze(1)  # (s, 1, 1, m)
+        dispatched_input = (dispatch_mask_expanded * reshaped_input_expanded).sum(dim=(0))  #  (s, m)
+
+        chunks = dispatched_input.chunk(self.num_experts, dim=0)
+        expert_outputs = []
+        for chunk, expert in zip(chunks, self.experts):
+            expert_outputs.append(expert(chunk))
+
+        expert_output = torch.cat(expert_outputs, dim=0)
+        # combined_output = torch.einsum("sec,ecm->sm", combine_weights.type_as(hidden_states), expert_output)
+        combine_exp = combine_weights.type_as(hidden_states).unsqueeze(3)  # (s, e, c, 1)
+        expert_exp = expert_output.unsqueeze(0)  # (1, e, c, m)
+        combined_output = (combine_exp * expert_exp).sum(dim=(1, 2))  # (s, m)
```

Comment:
well, this is rigurously equivalent to the approach we have in llama4 with scattering that uses the dispatch index, I don't think there is a difference with llama4 here no? let's standardize please